python/revng/scripting/projects/project.py

```diff
@@ -0,0 +1,240 @@
+        """
+        pass
+
+    def get_artifacts(self, params: Dict[str, List[StructBase]]) -> Dict[str, str | bytes]:
+        """
+        Allows fetching multiple artifacts at once. The `params` is a dict containing
+        the name of the artifact and a list of targets (it can be empty to fetch all the
+        targets).
+        Example `params`:
+        params = {
+            "disassemble": [Function_1, Function_2],
+            "decompile": []
+        }
+        """
+        result: Dict[str, str | bytes] = {}
+        for artifact_name, targets in params.items():
+            result[artifact_name] = self._get_artifact(artifact_name, targets)
+
+        return result
+
+    @abstractmethod
+    def analyze(
+        self, analysis_name: str, targets: Dict[str, List[str]] = {}, options: Dict[str, str] = {}
+    ):
+        """
+        Run a single analysis. In addition to the `analysis_name` you need to specify a dict
+        of targets, some analysis require you to also pass an `options` dict.
+        """
+        pass
+
+    @abstractmethod
+    def analyses_list(self, analysis_name: str):
+        """
+        Run analysis list, these are predefined list of analysis that run sequentially.
+        """
+        pass
+
+    def get_analysis_inputs(self, analysis_name: str) -> Dict[str, List[str]]:
+        """
+        Get the analysis container inputs and the associate acceptable kinds from
+        the `pipeline description`.
+        """
+        inputs: Dict[str, List[str]] = {}
+        for step in self.pipeline_description["Steps"]:
+            for analysis in step["Analyses"]:
+                if analysis["Name"] == analysis_name:
+                    for i in analysis["ContainerInputs"]:
+                        inputs[i["Name"]] = i["AcceptableKinds"]
+        return inputs
+
+    @abstractmethod
+    def commit(self):
+        """
+        Persist the changes from `self.model` to the backend.
+        """
+        pass
+
+    def revert(self):
+        """
+        Revert changes made to the `self.model` since the last call to `commit()`.
+        """
+        self._set_model(deepcopy(self.last_saved_model))
+
+    @abstractmethod
+    def _get_pipeline_description(self):
+        """
+        Define how to implement getting the pipeline description data, The method should parse the
+        data as `yaml` and store it in the variable `self.pipeline_description`.
+        """
+        pass
+
+    def import_and_analyze(self, input_binary_path: str):
+        """
+        A helper method for setting up the project. This method imports
+        the binary and runs the `revng` initial auto analysis.
+        """
+        self.import_binary(input_binary_path)
+        self.analyses_list("revng-initial-auto-analysis")
+        self._get_pipeline_description()
+        self._set_model_mixins()
+
+    def _set_revng_executable(self, revng_executable_path: str | None):
```

Comment:
The "`revng` executable" is something that makes sense only in a CLI profile, i.e., in `CLIProject`.
It doesn't belong to `Project`.

Reply:
I agree on a theoretical level, but `LocalDaemonProject` does require the revng executable

Reply:
That's his problem. `Project` should not assume in any way the `revng` executable is even available.

Reply:
Moved the `revng_executable` logic away into a mixin